basictracer/propagation.py

```diff
@@ -0,0 +1,105 @@
+from __future__ import absolute_import
+
+import base64
+import time
+import struct
+from opentracing import InvalidCarrierException, TraceCorruptedException
+from .context import Context
+from .span import BasicSpan
+from .wire_pb2 import TracerState
+from .util import generate_id
+
+_proto_size_bytes = 4 # bytes
+
+class BinaryPropagator(object):
+    def __init__(self, tracer):
+        self.tracer = tracer
+
+    def inject(self, span, carrier):
+        if type(carrier) is not bytearray:
+            raise InvalidCarrierException()
+        state = TracerState()
+        state.trace_id = span.context.trace_id
+        state.span_id = span.context.span_id
+        state.sampled = span.context.sampled
+        if span.context.baggage is not None:
+            for key in span.context.baggage:
+                state.baggage_items[key] = span.context.baggage[key]
+
+        # The binary format is {uint32}{protobuf} using big-endian for the uint
+        carrier.extend(struct.pack(">I", state.ByteSize()))
+        carrier.extend(state.SerializeToString())
+
+    def join(self, operation_name, carrier):
+        if type(carrier) is not bytearray:
+            raise InvalidCarrierException()
+        state = TracerState()
+        state.ParseFromString(str(carrier[_proto_size_bytes:]))
+        baggage = {}
```

Comment:
nit: IMO it would make more sense to maintain the invariant that `context.baggage` is never `None` (thus eliminating the check at line 25 above)...